torch/distributed/_sharded_tensor/api.py

```diff
@@ -751,3 +928,122 @@ def _create_tensor_from_params(*size, local_device, tensor_init_params: TensorIn
+        tensor([[1.+1.j, 2.+2.j], [1.+1.j, 2.+2.j]]) # Rank 0
+        tensor([[3.+3.j, 4.+4.j], [3.+3.j, 4.+4.j]]) # Rank 1
+        >>> partial_tensor = _PartialTensor(tensor)
+        >>> complete_tensor = partial_tensor.reshard(collect_spec)
+        >>> complete_tensor
+        tensor([4.+4.j, 6.+6.j]) # Rank 0
+        tensor([4.+4.j, 6.+6.j]) # Rank 1
+    """
+
+    def __init__(self, local_partial_shard, pg, aggr_op=distributed_c10d.ReduceOp.SUM):
+        self.local_partial_shard = local_partial_shard
+        self.pg = pg
+        self.aggr_op = aggr_op
+
+    def __post_init__(self):
+        if not isinstance(self.local_partial_shard, torch.Tensor):
+            raise ValueError("local_partial_shard needs to be a Tensor.")
+        if not isinstance(self.aggr_op, distributed_c10d.ReduceOp):
+            raise ValueError(
+                "aggr_op needs to be a member of distributed_c10d.ReduceOp."
+            )
+
+    def reshard(self, resharding_spec: ShardingSpec) -> ShardedTensor:
+        """
+        Consolidate all partial tensors across multiple ranks and covert to a sharded
+        tensor based on the resharding spec.
+
+        Args:
+            resharding_spec (:class:`torch.distributed._sharding_spec.ShardingSpec`):
+                The specification describing how we reshard the aggregated local result.
+
+        Returns:
+            A :class:`ShardedTensor` filled with local aggregated result.
+        """
+        if not isinstance(resharding_spec, ChunkShardingSpec):
+            raise NotImplementedError("Only ChunkShardingSpec supported for reshard.")
+        sharding_dim = int(resharding_spec.dim)  # type: ignore[attr-defined]
+        world_size = dist.get_world_size(self.pg)
+        local_shards = torch.tensor_split(self.local_partial_shard, world_size)
+        local_result = reduce_scatter(
+            torch.empty_like(local_shards[0]), list(local_shards), op=self.aggr_op
+        )
+
+        rank = dist.get_rank(self.pg)
+        local_metadata = None
+        shard_tensor_size = list(local_result.size())
+        shard_tensor_size[sharding_dim] *= world_size
+        current_offsets = [0] * len(local_result.size())
+        shards = []
+        for idx, placement in enumerate(resharding_spec.placements):  # type: ignore[attr-defined]
+            if rank == placement.rank():  # type: ignore[union-attr]
+                local_metadata = ShardMetadata(
+                    shard_offsets=current_offsets,
+                    shard_sizes=list(local_result.size()),
+                    placement=placement,
+                )
+                shards.append(Shard(local_result, local_metadata))
+                break
+            current_offsets[sharding_dim] += local_result.size(sharding_dim)  # type: ignore[index]
+
+        st = ShardedTensor._init_from_local_shards(
+            shards, tuple(shard_tensor_size), process_group=self.pg
+        )
+        st._sharding_spec = copy.deepcopy(resharding_spec)
```

Comment:
nit: why do we need to deepcopy the resharding_spec?

Reply:
Just on the safe side the pass-in spec might be changed later on. Are you concerned about the extra cost from deepcopy?

Reply:
yeah but I guess the cost is pretty small and it indeed avoid the potential of changing the pass-in spec.